index.rst

```diff
@@ -13,17 +13,256 @@
+
+For a client/server Butler the journal creation and deletion could happen in the client or in the server, but the former continues to permit a regular ``FileDatastore`` to be used.
+
+``Butler.import`` and ``Butler.transfer_from``
+""""""""""""""""""""""""""""""""""""""""""""""
+
+As with ``put``, we would write a journal file before the Datastore operations begin and delete it after Registry dataset writes succeed.
+Imports of dimension data, collections, and dataset types would happen before the journal file is written.
+``transfer_from`` (at least) would need to turn off its own journaling in favor of a higher-level one representing a QuantumGraph execution, when it is used in the transfer job for that execution.
+This suggests that we might want to make the public interfaces for transfer-from-graph and transfer-from-other-butler more distinct, and possibly make the former private.
+
+``Butler.get``
+""""""""""""""
+
+No journaling is needed, as this is a read-only operation.
+
+``Butler.unstore`` and ``Butler.purge``
+"""""""""""""""""""""""""""""""""""""""
+
+The journal file should be written before ``Registry`` transaction is committed and deleted only after all Datastore deletions succeed.
+This is slightly problematic for client/server, because the journal file will need to be populated with information we get from the Registry database; this means the client cannot be responsible for creating the journal file unless we make fetching the datastore records and deleting them separate operations.
+That isn't too bad - it's just a slight increase in latency and a bit more http traffic.
+Another alternative would be to have the server take responsibility for creating the journal file, and then either returning responsibility for its deletion to the client or taking responsibility for both the deletion of the Datastore artifacts and the deletion of the journal file.
+Which of these is preferable probably depends on whether we want these operations to block until completion and whether we have reasons to perform other Datastore operations on the server (up to this point, having the client use a vanilla ``FileDatastore`` and perform all Datastore operations still seems viable).
+
+.. note::
+   TODO: think about how all of this interacts with DECam raws.
+
+.. note::
+   TODO: think about how all of this affects disassembled composites.
+   Or get rid of them!  I am pretty confident now that we'll never need them, and dropping that before we embark on other major changes to Datastore seems wise (sorry, Tim).
+
+Public interface changes
+========================
+
+.. note::
+   It's all stubs and outlines from here on.
+
+Bundling DatasetRef with Datastore records
+------------------------------------------
+
+- If we're passing around DatasetRefs and bundles of datastore records together a lot, we should have a class for that.
```

Comment:
I'm guessing that @andy-slac has got very close to this in quantum graph. Since the system doesn't know anything until a specific datastore is created it has to be fairly generic. The "records" part of a `DatasetRef` would still have to be something like `dict[str, list[StoredDatastoreItemInfo]]` and also be a pydantic model for serialization purposes (the serialization would have to include the record subclass but I know how to do that).  `StoredDatastoreItemInfo` already has the UUID in it though so we have to be a bit careful -- there are cases where `list[StoredDatastoreItemInfo]` is fine (eg for `Butler.mexists()` calls) without having to include the hierarchy and memory bloat for `list[DatasetRef]` each carrying its own rows.

Reply:
From the perspective of integration with the query system, there are some advantages to actually putting the datastore records into the `DatasetRef` and saying that an "expanded" `DatasetRef` has both expanded data IDs (with dimension records) and datastore records.  I was playing around with that in some prototyping on this ticket last week, and I'm not sure its an idea that holds up everywhere.  But I think it's at least intuitive from a user perspective (I do not expect users to go poking at the Datastore records in a `DatasetRef`, at least most of the time, but I think they may care that a query they do include them, so they can use those refs in other places without forcing more lookups).